### PR TITLE
Make Solidity contracts upgradeable using UUPS proxies

### PR DESCRIPTION
**Pull request by GoodluckH**
I know that @bghughes had started working on adding hardhat #9, so please focus on changes made to the smart contracts only for now, and eventually we should merge with @bghughes's hardhat deployment/test implementations.

### CODE REVIEW

File: contracts/tokens/ArrowToken.sol [hunk 1]
```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+// Valid import paths for hardhat (need to comment out when using brownie)
+// import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+// import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+// import "@openzeppelin/contracts-upgradeable/access/AccessControlEnumerableUpgradeable.sol";
+// import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+// import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+// import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
```
Comment (BrassLion) [inline contracts/tokens/ArrowToken.sol]: Import paths can be made compatible with both Hardhat and Brownie, so there's no need to comment out import statements. The Brownie import substitution should instead be altered to match the NPM package names.

File: contracts/tokens/ArrowToken.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+contract ArrowToken is
+    Initializable,
+    ERC20Upgradeable,
+    AccessControlEnumerableUpgradeable,
+    OwnableUpgradeable,
+    UUPSUpgradeable
+{
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+    bytes32 public constant BURNER_ROLE = keccak256("BURNER_ROLE");
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
```
Comment (BrassLion) [inline contracts/tokens/ArrowToken.sol]: Adding access controls is out of scope for this PR, which is seeking to add upgradability to the token contract. We should try to keep PRs as small as possible and single-focused for simplicity.

I'm also not convinced we need complex roles for this contract beyond what is already available in the `Ownable` functionality. Since this contract will be managed by the Arrow multisig, there's no need for complex roles in my opinion.

File: contracts/tokens/ArrowToken.sol [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        address indexed to,
+        uint256 amount,
+        uint256 timestamp
+    );
+    event TokenBurn(
+        address indexed from,
+        address indexed account,
+        uint256 amount,
+        uint256 timestamp
+    );
```
Comment (BrassLion) [inline contracts/tokens/ArrowToken.sol]: These are also out of scope for the upgradable PR.

Additionally, minting and burning events already exist in the OpenZeppelin [ERC20 contract.](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L264)

File: contracts/tokens/ArrowToken.sol [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    // implement the UUPS interface
+    function _authorizeUpgrade(address) internal override onlyOwner {}
+
+    ///@notice Mint an `_initialSupply` amount of tokens upon contract creation
+    function initilize(uint256 _initialSupply) public initializer {
```
Comment (BrassLion) [inline contracts/tokens/ArrowToken.sol]:
```suggestion
    function initialize(uint256 _initialSupply) public initializer {
```

File: contracts/tokens/ArrowVestingBase.sol [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+// For brownie tests (should remove once we fully move to hardhat)
+import "@openzeppelin-upgradeable/contracts/finance/VestingWalletUpgradeable.sol";
+import "@openzeppelin-upgradeable/contracts/access/AccessControlEnumerableUpgradeable.sol";
+import "@openzeppelin-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import "@openzeppelin-upgradeable/contracts/proxy/utils/Initializable.sol";
+import "@openzeppelin-upgradeable/contracts/proxy/utils/UUPSUpgradeable.sol";
+
+/// @title Base implementation of vesting contract that will be cloned into individual vesting wallets pointing at specific beneficiary addresses.
+contract ArrowVestingBase is
```
Comment (BrassLion) [inline contracts/tokens/ArrowVestingBase.sol]: I don't see why this contract needs to be upgradable. The OpenZeppelin contract already has all the functionality we need, and we can always deploy a new `ArrowVestingFactory` contract in the very worst case.

**Review by BrassLion — CHANGES_REQUESTED** [review]
Thanks for looking at this. 

As a general rule, we should keep changes in each PR to a minimum, and only implement one change, creating multiple PRs if necessary.

As such, please don't relocate the contracts into the `tokens` directory along with these changes, it messes with the git history. That should it's own PR if we expand the types of contracts we implement.

The upgradablable changes look good. We really need extensive unit testing on that functionality before this change could be merged, to make sure everything does what we expect, since this change introduces a lot more complexity to the token contract.

**Comment by GoodluckH** [comment]
Thank you for these comments, @BrassLion. I will reflect your suggestions and edits in the next commit.

 As for testing, brownie just doesn't work on my device, despite hours of effort troubleshooting it (my python envs and paths are pretty messed up). I was wondering if I could build a hardhat-based test suite for the next commit under the same PR? I think it's still within the scope since testing is a necessary part of the process.

File: contracts/tokens/ArrowVestingBase.sol (hunk 5 above)
Comment (GoodluckH) [reply contracts/tokens/ArrowVestingBase.sol:19]: I thought this contract is already upgradeable? https://github.com/Arrow-air/Arrow-Contracts/blob/main/contracts/ArrowVestingBase.sol

All I changed was to make it conform to UUPS proxies.

Comment (GoodluckH) [reply contracts/tokens/ArrowVestingBase.sol]: Ahh, I see your point.  Will resolve this.

**Comment by GoodluckH** [comment]
I think the nodejs CI/CD flow has been properly configured and was able to run all tests. @BrassLion Take a look at the new test suite -- it's basically a rewrite of those brownie tests, and I added a module to test upgradeable contracts.

**Comment by bghughes** [comment]
> I know that @bghughes had started working on adding hardhat #9, so please focus on changes made to the smart contracts only for now, and eventually we should merge with @bghughes's hardhat deployment/test implementations.

Sounds great - btw, I agree on the move to UUPS proxies in practice though they are less tested at market

File: .github/workflows/node.js.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions
```
Comment (BrassLion) [inline .github/workflows/node.js.yml]: Try not to include old code from the examples you adapt if possible. This is no longer an accurate comment.

File: .github/workflows/node.js.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+name: Node.js CI
```
Comment (BrassLion) [inline .github/workflows/node.js.yml]: This could be named more accurately. Something like `Hardhat Tests`?

File: README.md [hunk 8]
```diff
@@ -1,27 +1,13 @@
-# Arrow Contracts
-
-Implementation of the Arrow token and governance smart contracts.
+# Arrow Contract
```
Comment (BrassLion) [inline README.md]: Why are we changing the README to remove information?

File: README.md [hunk 9]
```diff
@@ -1,27 +1,13 @@
...
-## Deployments
+Clone the repo and then run `npm install` to get started.
 
-### Rinkeby
+## Deploy contracts
 
-| Contract | Address |
-| -- | -- |
-| Gnosis Safe | 0x6943eBEfCD7d85B536aEb35BBFd95C5699158Abe |
+[TODO]
```
Comment (BrassLion) [inline README.md:9]: We shouldn't remove deployed contract addresses. These exist on the testnet right now, and haven't been updated.

File: README.md [hunk 10]
```diff
@@ -1,27 +1,13 @@
...
-
-## Deployments
+Clone the repo and then run `npm install` to get started.
 
-### Rinkeby
+## Deploy contracts
 
-| Contract | Address |
-| -- | -- |
-| Gnosis Safe | 0x6943eBEfCD7d85B536aEb35BBFd95C5699158Abe |
```
Comment (BrassLion) [inline README.md:23]: We shouldn't remove contract addresses for existing contracts. These already exist on the testnet and haven't been updated.

File: README.md [hunk 11]
```diff
@@ -1,27 +1,13 @@
...
+[TODO]
 
-## License
+## Test
 
-This project is licensed under the [MIT license](LICENSE).
```
Comment (BrassLion) [inline README.md]: We shouldn't remove the license section.

File: README.md (hunk 9 above)
Comment (BrassLion) [inline README.md]: We shouldn't remove deployed contract addresses. These exist on the testnet right now, and haven't been updated.

File: contracts/ArrowToken.sol [hunk 12]
```diff
@@ -1,27 +1,36 @@
...
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
 
-import "@openzeppelin/contracts/access/Ownable.sol";
-import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+/// @title Arrow Token implementation that is upgradeable
+/// @notice Create ARROW tokens of an `initialSupply` amount
+/// @dev If we were to upgrade the initializer by introducing new variables, we need to create a normal function like `initializeV2` (https://forum.openzeppelin.com/t/uups-proxies-tutorial-solidity-javascript/7786/53), and make sure to include all existing variables from the current version of the contract to avoid storage collision
```
Comment (BrassLion) [inline contracts/ArrowToken.sol]: Please follow the existing style for comments, so we're not mixing different styles in the codebase.
```suggestion
/**
    @title Arrow Token implementation that is upgradeable
    @notice Create ARROW tokens of an `initialSupply` amount
    @dev If we were to upgrade the initializer by introducing new variables, we need to create a normal function like `initializeV2` (https://forum.openzeppelin.com/t/uups-proxies-tutorial-solidity-javascript/7786/53), and make sure to include all existing variables from the current version of the contract to avoid storage collision
*/
```

File: contracts/ArrowToken.sol [hunk 13]
```diff
@@ -1,27 +1,36 @@
...
 
-        Only accepts calls from the contract owner.
-     */
-    function mint(address to, uint256 amount) public onlyOwner
-    {
-        _mint(to, amount);
+    /// @notice Mint and issue new tokens to a specified address
+    /// @param _to the address to issue new tokens to
+    /// @param _amount the amount of tokens to be mint and transfered to `_to`
+    function issueNewTokens(address _to, uint256 _amount) public onlyOwner {
```
Comment (BrassLion) [inline contracts/ArrowToken.sol]: Why are we changing the function name here? It's a non-standard method name I haven't seen before.

File: contracts/ArrowVestingBase.sol [hunk 14]
```diff
@@ -1,21 +1,22 @@
...
+        address _beneficiaryAddress,
+        uint64 _startTimestamp,
+        uint64 _durationSeconds
     ) public initializer {
-        __VestingWallet_init(beneficiaryAddress, startTimestamp, durationSeconds);
-    } 
+        __VestingWallet_init(
+            _beneficiaryAddress,
+            _startTimestamp,
+            _durationSeconds
```
Comment (BrassLion) [inline contracts/ArrowVestingBase.sol]: Let's keep everything in this PR relevant to upgradable contracts. We can open up another PR to fix variable name styling if necessary later to keep everything compartmentalised.

**Review by BrassLion — DISMISSED** [review]
Thanks for taking a look at migrating the test code over to Hardhat.

We should really be making PRs as focused as possible, in order to make reviewing changes easy. 

This PR should focus on the minimum code changes required to make the token contract upgradable, with associated tests. Any other code should be moved to a new PR so those changes can be added incrementally.

I'd suggest watching #9 closely, and maybe building on top of that branch so you can avoid any conflicts with it, since it will need to be merged first.

File: README.md (hunk 8 above)
Comment (GoodluckH) [reply README.md]: This README was auto-generated by hardhat, I will change it back to the original version with added instructions on running tests in hardhat.

**Comment by GoodluckH** [comment]
> 



> Thanks for taking a look at migrating the test code over to Hardhat.
> 
> We should really be making PRs as focused as possible, in order to make reviewing changes easy.
> 
> This PR should focus on the minimum code changes required to make the token contract upgradable, with associated tests. Any other code should be moved to a new PR so those changes can be added incrementally.
> 
> I'd suggest watching #9 closely, and maybe building on top of that branch so you can avoid any conflicts with it, since it will need to be merged first.

@BrassLion Thank you for the feedback. I have resolved your comments and this PR should be good to go.

We should really discuss and decide on what to do next because #9 is developing scripts based on un-upgradable contracts. If we want to keep each PR to its scope, then I cannot really change #9's contracts to upgradeable and thereby build on top of that branch. Let me know what you think is the best course of action to take from here.

**Comment by bghughes** [comment]
Just a heads up @GoodluckH - it looks like you exposed this Admin Key in the `.env` publicly

**Comment by GoodluckH** [comment]
> Just a heads up @GoodluckH - it looks like you exposed this Admin Key in the `.env` publicly

@bghughes Hmm, I didn't really have an `.env` for this PR, maybe that's your env lol.
